freezing w aot

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105497

Freezing will take parameters and turn them into constants. A couple changes here: 

-  move the setting of `flat_params[dropped_index]` before cpp compilation so that cpp_wrapper knows they have been dropped
- compile_fx_aot is doesn't use aot_autograd for invocation, so we no longer add the wrapper which discards dropped param indices. Continuing to add arguments everywhere didn't seem great, so I added `_in_aot_compilation`, but maybe reviewers would prefer something else. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

eellison committed:
freezing w aot

[ghstack-poisoned]

eellison committed:
Update on "freezing w aot"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "freezing w aot"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on test/inductor/test_inductor_freezing.py:
```diff
@@ -364,6 +364,21 @@ def foo(mod, inp):
+
+        out_eager = mod(x)
+
+        with torch.no_grad():
```
I added a test for cpp_wrapper but I"m not sure how I should exercise the export path. 
If I do something like [this](https://github.com/pytorch/pytorch/blob/main/test/cpp/aot_inductor/test.py#L27) do I need to do any cleanup ?

desertfire replied:
To test Export+AOTInductor, you can use a benchmarking script with --export-aot-inductor. alexnet should just do it.

I am also adding a unit test file in my another PR. You can also add a test to that later.

desertfire reviewed (commented):
Make sense to me, but let's test it out.

desertfire commented on torch/_inductor/compile_fx.py:
```diff
@@ -771,12 +778,13 @@ def compile_fx_aot(
         if config_patches is None
         else {**config_patches, "cpp_wrapper": True}
```
You can set `config.freezing` to True here.

eellison replied:
That would turn it always on. Just confirming that's what we want

desertfire replied:
Yes. Later when when we have the weight as a separate file solution ready, we can drop it.

eellison replied:
okay, then we don't need the separate test if it's always enabled

eellison replied:
Right ?

desertfire replied:
https://github.com/pytorch/pytorch/pull/105496/files , I have added a test_aot_inductor.py. It will be nice if you can add a test there to test a model with weights. You will need to change how we pass in args to compiled model there.

eellison committed:
Update on "freezing w aot"


Freezing will take parameters and turn them into constants. A couple changes here: 

-  move the setting of `flat_params[dropped_index]` before cpp compilation so that cpp_wrapper knows they have been dropped
- compile_fx_aot is doesn't use aot_autograd for invocation, so we no longer add the wrapper which discards dropped param indices. Continuing to add arguments everywhere didn't seem great, so I added `_in_aot_compilation`, but maybe reviewers would prefer something else. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "freezing w aot"


Freezing will take parameters and turn them into constants. A couple changes here: 

-  move the setting of `flat_params[dropped_index]` before cpp compilation so that cpp_wrapper knows they have been dropped
- compile_fx_aot is doesn't use aot_autograd for invocation, so we no longer add the wrapper which discards dropped param indices. Continuing to add arguments everywhere didn't seem great, so I added `_in_aot_compilation`, but maybe reviewers would prefer something else. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "freezing w aot"


Freezing will take parameters and turn them into constants. A couple changes here: 

-  move the setting of `flat_params[dropped_index]` before cpp compilation so that cpp_wrapper knows they have been dropped
- compile_fx_aot is doesn't use aot_autograd for invocation, so we no longer add the wrapper which discards dropped param indices. Continuing to add arguments everywhere didn't seem great, so I added `_in_aot_compilation`, but maybe reviewers would prefer something else. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "freezing w aot"


Freezing will take parameters and turn them into constants. A couple changes here: 

-  move the setting of `flat_params[dropped_index]` before cpp compilation so that cpp_wrapper knows they have been dropped
- compile_fx_aot is doesn't use aot_autograd for invocation, so we no longer add the wrapper which discards dropped param indices. Continuing to add arguments everywhere didn't seem great, so I added `_in_aot_compilation`, but maybe reviewers would prefer something else. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "unrelated failed test"
